src/main/java/org/spongepowered/common/util/SpongeColor.java

```diff
@@ -0,0 +1,184 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.util;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import org.apache.commons.lang3.Validate;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.DataView;
+import org.spongepowered.api.data.Queries;
+import org.spongepowered.api.data.persistence.AbstractDataBuilder;
+import org.spongepowered.api.data.persistence.InvalidDataException;
+import org.spongepowered.api.util.Color;
+
+import java.util.Optional;
+
+public final class SpongeColor implements Color {
+
+    private final int rgb;
+    private final int red;
+    private final int green;
+    private final int blue;
+
+    private SpongeColor(int red, int green, int blue) {
+        this.rgb = red << 16 | green << 8 | blue;
+        this.red = red;
+        this.green = green;
+        this.blue = blue;
+    }
+
+    @Override
+    public int getRgb() {
+        return this.rgb;
+    }
+
+    @Override
+    public int getRed() {
+        return this.red;
+    }
+
+    @Override
+    public int getGreen() {
+        return this.green;
+    }
+
+    @Override
+    public int getBlue() {
+        return this.blue;
+    }
+
+    @Override
+    public int getContentVersion() {
+        return 1;
+    }
+
+    @Override
+    public DataContainer toContainer() {
+        return DataContainer.createNew()
+                .set(Queries.CONTENT_VERSION, this.getContentVersion())
+                .set(Queries.COLOR_RED, this.getRed())
+                .set(Queries.COLOR_GREEN, this.getGreen())
+                .set(Queries.COLOR_BLUE, this.getBlue());
+    }
+
+    @Override
+    public int hashCode() {
+        return rgb;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return this == obj || obj != null && getClass() == obj.getClass() && hashCode() == obj.hashCode();
+    }
+
+    @Override
+    public String toString() {
+        return com.google.common.base.MoreObjects.toStringHelper(this)
+                .add("rgb", this.rgb)
+                .add("red", this.red)
+                .add("green", this.green)
+                .add("blue", this.blue)
+                .toString();
+    }
+
+    public static final class Builder extends AbstractDataBuilder<Color> implements Color.Builder {
+
+        private int red = -1;
+        private int green = -1;
+        private int blue = -1;
+
+        public Builder() {
+            super(Color.class, 1);
+        }
+
+        @Override
+        public Builder rgb(int rgb) {
+            return rgb(rgb >> 16, rgb >> 8, rgb);
+        }
+
+        @Override
+        public Builder rgb(int red, int green, int blue) {
+            this.red = red & 0xFF;
+            this.green = green & 0xFF;
+            this.blue = blue & 0xFF;
+            return this;
+        }
+
+        @Override
+        public Builder mix(Color... colors) {
```

Comment:
I'm shaky about this behaviour. A builder shouldn't do any processing like this pre-build.

My feeling is that we should offer a `mix(Color)` method that we can chain up (and this varargs method would default to chaining these up), and that we store the colours that we want to mix. Then, when `build()` is called, that's when the mixing occurs. This means that we can use the builder pattern for both mixing colours and specifiying a colour from RGB values, or both.

It allows us to offer a `Color.mix(Color...)` helper method that just calls `Color.builder().mix(...).build()` too.

To emulate the current behaviour, you could then just create the colour in a separate builder, so `Color.build().rgb(...).mix(Color.mix(...)).build()`

Reply:
I don't think I'm understanding what you mean by this. I believe you're suggesting that each call to `mix` should save any passed in colors to a list, which would be mixed internally when build is called.

I'm still a proponent of deprecated the mix methods entirely as per my reasoning in the API pull. One of the things for me is that the `Color` class, in my opinion, should not be represented by a builder - it's only like that so the implementation can exist in Common and use the builder registry in API. The methods don't interact in the same way as most builders as ever method essentially overrides another.

With the change to reset to have the default values be `-1`, the way this method *should* be functioning is to mix with an existing color in the builder if defined. I believe what you're getting at is that chaining the methods together should have the same effect as combining them all together into one call, which they currently do not.

However, I would still prefer deprecating these methods for removal as they are impractical.